src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,274 @@
+
+        // This should be set before flushing the buffer.
+        public TdsStreamPacketType PacketHeaderType { get; set; }
+
+        internal byte PacketNumber { get; private set; } = 1; // Packets always start with 1.
+
+        public TdsWriteStream(Stream underLyingStream) : base()
+        {
+            _underlyingStream = underLyingStream;
+            _WriteBuffer = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+        }
+
+        /// <summary>
+        /// Should only be called after the login negotiation is done.
+        /// </summary>
+        /// <param name="bufferSize">new buffer size</param>
+        public void SetPacketSize(int bufferSize) => _WriteBuffer = new byte[bufferSize];
+        
+        public void ReplaceUnderlyingStream(Stream stream)
+            => _underlyingStream = stream;
+
+        /// <inheritdoc />
+        public override bool CanRead => false;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <summary>
+        /// Always returns true since we can always write to the stream.
+        /// </summary>
+        public override bool CanWrite => true;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override async Task FlushAsync(CancellationToken ct)
+        {
+            await FlushAsync(ct, true).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Use hard flush when we definitely know that the end of packet is reached.
+        /// </summary>
+        /// <param name="ct">Cancellation token for async operation.</param>
+        /// <param name="isAsync">Bool to indicate async operation</param>
+        /// <param name="flushMode">Whether this is a hard flush or a softflush</param>
+        private async Task FlushAsync(CancellationToken ct, bool isAsync, FlushMode flushMode = FlushMode.SoftFlush)
+        {
+            Debug.Assert(PacketHeaderType != TdsStreamPacketType.None, "PacketHeaderType is not set. Cannot flush the buffer without setting the packet header type.");
+            _WriteBuffer[0] = (byte)PacketHeaderType;
+            byte status;
+
+            // TODO: Handle cancellation queueing. If there is a cancellation queued up, then send the status with IGNORE bit set.
+
+            if (flushMode == FlushMode.HardFlush)
+            {
+                status = TdsEnums.ST_EOM;
+            }
+            else
+            {
+                status = TdsEnums.ST_BATCH;
+                // We have filled up the buffer. We need to send it out, hence next packet gets an increment.
+                PacketNumber++;
+            }
+
+            _WriteBuffer[1] = status;
+            _WriteBuffer[2] = (byte)(WriteBufferOffset >> 8);   // Length upper byte
```

Comment:
Two micro-optimisations (since this is on the hot path.) Also, one nit.

Nit: elements 4, 5 don't need to be set in `FlushAsync` - they're constant. We can set them once on buffer initialisation.

Nit: BinaryPrimitives will provide a negligible performance boost when writing the packet length.

Nit: I'm not a fan of using `WriteBufferOffset` as the length. I'd personally prefer to see it described as CurrentPacketLength (or something clearer) throughout.

Reply:
+1 to `BinaryPrimitives` (will need to pull in a nuget package for netfx tho)

Reply:
@edwardneal  and @benrr101 Ack. Will address them. 

@benrr101  on a separate note, this is one of the reason I am hoping to keep Read and Write Streams separate. When the read stream comes in the packetLength is the length needed from the packet header, which plays an important role. 

I am really hoping to keep the logic of read and write paths in separate classes. however there may be a better way of dealing with my conundrum, which you guys can help with, when you see my Read Stream PR